nullable variable docs (#29845)
hashicorp/terraform: website/docs/language/values/variables.html.md

random_pet.animal: Creation complete after 0s [id=jae-known-mongoose]
```

### Disallowing null Module Input Values

[inpage-nullable]: #disallowing-null-input-values

-> This feature was introduced in Terraform v1.1.0.

> The default value for `nullable` is `true`. When `nullable` is `true`, `null`

Reviewer: I think it'd be good to have an introductory paragraph here that allows the reader to know we're about to talk about an argument called `nullable`, before getting into the details:

> The `nullable` argument in a `variable` block controls whether the module caller may assign the value `null` to the variable:
>
> ```hcl
> variable "example" {
>   type     = string
>   nullable = false 
> }
> ```

...and then from there keep these existing paragraphs, starting with "The default value for `nullable` is ...", so that hopefully by the time we're reading that we understand what exactly we mean by `nullable` in this sentence.

Reviewer 2: For what it's worth, I agree with @apparentlymart here! That's a great suggestion!

is a valid value for the variable, and the module configuration must always
account for the possibility of the variable value being `null`. Passing a
`null` value as a module input argument will override any `default` value.

Setting `nullable` to `false` ensures that the variable value will never be
`null` within the module. If `nullable` is `false` and the variable has a
`default` value, then the default will be used when a `null` value is given as
a module input argument.

> 

Reviewer: Perhaps we should note here that the restriction applies only to the top-level value, and not to any nested values. Something like this?

> `nullable` only controls where the direct value of the variable may be `null`. For variables of collection or structural types, such as list or object types, the caller may still use `null` in nested elements or attributes.

random_pet.animal: Creation complete after 0s [id=jae-known-mongoose]
```

### Disallowing null Module Input Values

[inpage-nullable]: #disallowing-null-input-values

> -> This feature was introduced in Terraform v1.1.0.

Reviewer 2: ```suggestion
-> This feature is available in Terraform v1.1.0 and later.
```

> The `nullable` argument in a variable block controls whether the module caller
> may assign the value `null` to the variable:

Reviewer 2: ```suggestion
The `nullable` argument in a variable block controls whether the module caller
may assign the value `null` to the variable.
```

```
variable "example" {
  type     = string
  nullable = false 
}
```

The default value for `nullable` is `true`. When `nullable` is `true`, `null`
is a valid value for the variable, and the module configuration must always
account for the possibility of the variable value being `null`. Passing a
`null` value as a module input argument will override any `default` value.

> Setting `nullable` to `false` ensures that the variable value will never be
> `null` within the module. If `nullable` is `false` and the variable has a
> `default` value, then the default will be used when a `null` value is given as
> a module input argument.

Reviewer 2: ```suggestion
Setting `nullable` to `false` ensures that the variable value will never be
`null` within the module. If `nullable` is `false` and the variable has a
`default` value, then Terraform uses the default when a module input argument is `null`.
```

> `nullable` only controls where the direct value of the variable may be `null`.
> For variables of collection or structural types, such as list or object types,
> the caller may still use `null` in nested elements or attributes, as long as
> the collection or structure itself is not null.

Reviewer 2: ```suggestion
The `nullable` argument only controls where the direct value of the variable may be `null`.
For variables of collection or structural types, such as lists or objects,
the caller may still use `null` in nested elements or attributes, as long as
the collection or structure itself is not null.
```
